File: Streaming Data Analysis/flajolet-martin algorithm.scala
// package com.datamining.hw5

import java.io.{File, FileOutputStream, PrintWriter}
import java.text.SimpleDateFormat
import java.util
import java.util.Calendar
import org.apache.spark.{SparkConf, streaming}
import org.apache.spark.streaming.{Seconds, StreamingContext}
import org.json4s.{DefaultFormats, JObject}
import org.json4s.jackson.JsonMethods.parse
import scala.math.pow
import java.io.FileWriter



object swathi_nayak_task2 {


  def convertJsonToState(jsonStr: String): String = {
    implicit val jsonFormat = DefaultFormats
    val key = parse(jsonStr).asInstanceOf[JObject]
    val bus_state = (key \ "city").extract[String]
    bus_state
  }

  def main(args: Array[String]): Unit = {

    val conf: SparkConf = new SparkConf().setAppName("swathi_nayak_task2").setMaster("local")

    val port = args(0)
    val output_file = args(1)

    val out_file = new PrintWriter(new File(output_file))
    val output_header = "Time,Ground Truth,Estimation\n"
    out_file.write(output_header)
    out_file.close()

    val ssc = new StreamingContext(conf, Seconds(5))
    ssc.sparkContext.setLogLevel(logLevel = "OFF")
    val substream = ssc.socketTextStream("localhost", port.toInt)
    val wd = streaming.Duration(30000)
    val sd = streaming.Duration(10000)
    val streaming_window = substream.window(windowDuration = wd, slideDuration = sd)

    val n=12
    var len_trailing_zero: Array[Int] = new Array(n)
    var max_trailing_zero: Array[Int] = new Array(n)
    var city_set : util.HashSet[String] = new util.HashSet[String]()

    streaming_window.repartition(1)
    streaming_window.foreachRDD(myrdd => {myrdd.repartition(1)
      var chunks: Array[String] = myrdd.collect()
      for(chunk <- chunks){
      val bus_city = convertJsonToState(chunk)
      val bus_int = bus_city.hashCode
      val mod_vals = Array(17, 83, 29, 253, 97, 173, 53, 12, 37, 317, 79, 11)

      var bus_hash: Array[Int] = new Array(n)
      var bus_binary: Array[String] = new Array(n)

      for (i <- 0 to 11) {
        bus_hash(i) = bus_int % mod_vals(i)
      }

      for (i <- 0 to 11) {
        bus_binary(i) = bus_hash(i).toBinaryString
      }


      for (i <- 0 to 11) {
        val temp = bus_binary(i).reverse.dropWhile(_ == '0').reverse
        if ( bus_binary(i).length != (bus_binary(i).length - temp.length)) {
          len_trailing_zero(i) = bus_binary(i).length - temp.length
        }
      }

      for (i <- 0 to 11) {
        if (len_trailing_zero(i) > max_trailing_zero(i)) {
          max_trailing_zero(i) = len_trailing_zero(i)
        }
      }

      city_set.add(bus_city)

    }
      var grouped_averages: Array[Int] = new Array(6)

      for (i <- 0 until 11 by 2){
        grouped_averages(i/2) = (pow(2, max_trailing_zero(i)).toInt + pow(2, max_trailing_zero(i+1)).toInt)/2

      }

      grouped_averages = grouped_averages.sorted
      val estimate = grouped_averages(3).toString


      val fw = new PrintWriter(new FileOutputStream(new File(output_file),true))
      // val fw = new FileWriter(output_file, true)
      val format = new SimpleDateFormat("y-M-d h:m:s")
      val currdate = format.format(Calendar.getInstance().getTime())
      fw.write(currdate + "," + city_set.size().toString + "," + estimate + "\n")
      // fw.write(city_set.size().toString + "," + estimate + "\n")
      fw.close()
      // out_file.write(currdate + "," + city_set.size().toString + "," + estimate + "\n")

      })

    ssc.start()
    ssc.awaitTermination()

  }
}

File: Streaming Data Analysis/flajolet-martin algorithm.py
from pyspark import SparkContext
from pyspark.streaming import StreamingContext
import binascii
import datetime
import json
import sys


def hash_functions(x):
	m=671
	return [(7*x+1)%m, (3*x+1)%m, (51*x+58)%m, (68*x+73)%m, (17*x+19)%m, (42*x+78)%m, (17*x+19)%m, (136*x+152)%m, (23*x+29)%m, (13*x+19)%m, (11*x+19)%m, (23*x+37)%m]


def implement_flajolet_martin(stream_rdd):

	n = 12
	len_trailing_zero = [0] * n
	max_trailing_zero = [0] * n

	distinct_city = dict()

	data_stream = stream_rdd.collect()

	for business_data in data_stream:

		business = json.loads(business_data)
		bus_city = business['city']
		b_hash = int(binascii.hexlify(bus_city.encode('utf8')), 16)
		mod_vals = [17, 83, 29, 253, 97, 173, 53, 12, 37, 317, 79, 11]
		bus_hash = []
		bus_binary = []

		for i in range(n):
			bus_hash.append(b_hash % mod_vals[i])

		for i in range(n):
			bus_binary.append(format(bus_hash[i], '016b'))

		for i in range(n):
			if len(bus_binary[i]) != len(bus_binary[i])-len(bus_binary[i].rstrip('0')):
				len_trailing_zero[i] = len(bus_binary[i])-len(bus_binary[i].rstrip('0'))

		for i in range(n):
			if len_trailing_zero[i] > max_trailing_zero[i]:
				max_trailing_zero[i] = len_trailing_zero[i]

		distinct_city[bus_city] = 1

	grouped_averages = []
	for i in range(0, n, 2):
		grouped_averages.append((2**max_trailing_zero[i]+2**max_trailing_zero[i+1])//2)
	grouped_averages = sorted(grouped_averages)
	estimate = str(grouped_averages[3])

	f = open(output_file, "a+")
	f.write(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S") + "," + str(len(distinct_city)) + "," + estimate + "\n")
	f.close()


if __name__ == "__main__":
	port = int(sys.argv[1])
	output_file = sys.argv[2]

	sc = SparkContext()
	sc.setLogLevel(logLevel="ERROR")

	f = open(output_file, "w+")
	f.write("Time,Ground Truth,Estimation\n")
	f.close()

	stream_context = StreamingContext(sc, 5)
	stream_ctxt = stream_context.socketTextStream("localhost", port)
	streaming_window = stream_ctxt.window(30, 10)
	streaming_window.foreachRDD(implement_flajolet_martin)
	stream_context.start()
	stream_context.awaitTermination()
